In [1]:
import pandas as pd

df=pd.read_csv('Smart_Outcome_Predictor_Dataset_5200.csv')
df.head(5)

,student_id,age,country_region,device_type,education_background,course_level,course_category,course_start_date,week_of_year,sessions,time_spent_hours,videos_watched,quiz_attempts,assignments_submitted,forum_posts,avg_quiz_score,attendance_rate,completion_status,final_score
0,700001,32,Europe,Laptop,Undergrad,Intermediate,Business,2024-03-18,12,1,7.6,1,6,1,1,53.3,0.655,0,49.8
1,700002,17,Europe,Laptop,Undergrad,Intermediate,Programming,2024-08-22,34,16,27.2,6,4,7,1,51.5,1.000,1,84.0
2,700003,25,Europe,Mobile,Graduate,Advanced,Programming,2024-09-28,39,6,7.1,16,2,2,0,62.2,0.810,0,62.5
3,700004,26,Asia,Mobile,Undergrad,Beginner,Design,2024-03-09,10,34,22.1,57,9,6,0,59.3,0.875,1,89.5
4,700005,26,Asia,Tablet,WorkingPro,Advanced,Business,2024-03-21,12,22,32.3,41,9,2,0,65.1,0.814,0,67.4


In [2]:
df.isnull().sum()

student_id                 0
age                        0
country_region             0
device_type                0
education_background       0
course_level               0
course_category            0
course_start_date          0
week_of_year               0
sessions                   0
time_spent_hours         112
videos_watched             0
quiz_attempts              0
assignments_submitted      0
forum_posts                0
avg_quiz_score            81
attendance_rate           80
completion_status          0
final_score                0
dtype: int64

Fill missing values

In [3]:
df["time_spent_hours"] = df["time_spent_hours"].fillna(df["time_spent_hours"].median())
df["avg_quiz_score"] = df["avg_quiz_score"].fillna(df["avg_quiz_score"].median())
df["attendance_rate"] = df["attendance_rate"].fillna(df["attendance_rate"].median())

print(df.isnull().sum())

student_id               0
age                      0
country_region           0
device_type              0
education_background     0
course_level             0
course_category          0
course_start_date        0
week_of_year             0
sessions                 0
time_spent_hours         0
videos_watched           0
quiz_attempts            0
assignments_submitted    0
forum_posts              0
avg_quiz_score           0
attendance_rate          0
completion_status        0
final_score              0
dtype: int64


* Part B: Data understanding & preparation

Identify features and target variables for both tasks.

* Features - student_id,age,week_of_year,sessions,time_spent_hours,videos_watched,quiz_attempts,assignments_submitted,forum_posts,avg_quiz_score,attendance_rate,
country_region,device_type,education_background,course_level,course_category,course_start_date

* Target - completion_status & final_score

In [4]:
X = df.drop(columns=["completion_status", "final_score"])
y_classification = df["completion_status"]
y_regression = df["final_score"]

print('Features Shape:', X.shape)
print('Classification Target Shape:', y_classification.shape)
print('Regression Target Shape:', y_regression.shape)

Features Shape: (5200, 17)
Classification Target Shape: (5200,)
Regression Target Shape: (5200,)


Perform a train-test split for classification and regression problems.

In [5]:
from sklearn.model_selection import train_test_split

X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(
    X,y_classification,test_size=0.2,random_state=42,stratify=y_classification)

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X,y_regression,test_size=0.2,random_state=42)

print('Classification Training Shape:', X_train_cls.shape)
print('Classification Testing Shape:', X_test_cls.shape)
print('Regression Training Shape:', X_train_reg.shape)
print('Regression Testing Shape:', X_test_reg.shape)

Classification Training Shape: (4160, 17)
Classification Testing Shape: (1040, 17)
Regression Training Shape: (4160, 17)
Regression Testing Shape: (1040, 17)


Apply basic preprocessing (scaling/encoding if required).

In [6]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler

numerical_cols = ['age','week_of_year','sessions','time_spent_hours','videos_watched','quiz_attempts','assignments_submitted','forum_posts','avg_quiz_score','attendance_rate']
categorical_cols = ['country_region','device_type','education_background','course_level','course_category']

scaler = StandardScaler()

X_train_num_cls = pd.DataFrame(
    scaler.fit_transform(X_train_cls[numerical_cols]),
    columns=numerical_cols,
    index=X_train_cls.index
)

X_test_num_cls = pd.DataFrame(
    scaler.transform(X_test_cls[numerical_cols]),
    columns=numerical_cols,
    index=X_test_cls.index
)

encoder = OneHotEncoder(handle_unknown="ignore", sparse_output=False)

X_train_cat_cls = pd.DataFrame(
    encoder.fit_transform(X_train_cls[categorical_cols]),
    columns=encoder.get_feature_names_out(categorical_cols),
    index=X_train_cls.index
)

X_test_cat_cls = pd.DataFrame(
    encoder.transform(X_test_cls[categorical_cols]),
    columns=encoder.get_feature_names_out(categorical_cols),
    index=X_test_cls.index
)

X_train_cls = pd.concat([X_train_num_cls, X_train_cat_cls], axis=1)
X_test_cls = pd.concat([X_test_num_cls, X_test_cat_cls], axis=1)

scaler = StandardScaler()

X_train_num_reg = pd.DataFrame(
    scaler.fit_transform(X_train_reg[numerical_cols]),
    columns=numerical_cols,
    index=X_train_reg.index
)

X_test_num_reg = pd.DataFrame(
    scaler.transform(X_test_reg[numerical_cols]),
    columns=numerical_cols,
    index=X_test_reg.index
)

encoder = OneHotEncoder(handle_unknown="ignore", sparse_output=False)

X_train_cat_reg = pd.DataFrame(
    encoder.fit_transform(X_train_reg[categorical_cols]),
    columns=encoder.get_feature_names_out(categorical_cols),
    index=X_train_reg.index
)

X_test_cat_reg = pd.DataFrame(
    encoder.transform(X_test_reg[categorical_cols]),
    columns=encoder.get_feature_names_out(categorical_cols),
    index=X_test_reg.index
)

X_train_reg = pd.concat([X_train_num_reg, X_train_cat_reg], axis=1)
X_test_reg = pd.concat([X_test_num_reg, X_test_cat_reg], axis=1)
X_train_cls.head()

,age,week_of_year,sessions,time_spent_hours,videos_watched,quiz_attempts,assignments_submitted,forum_posts,avg_quiz_score,attendance_rate,...,education_background_Undergrad,education_background_WorkingPro,course_level_Advanced,course_level_Beginner,course_level_Intermediate,course_category_Business,course_category_Data,course_category_Design,course_category_Marketing,course_category_Programming
849,-0.422862,0.813102,-0.384893,-0.194625,-0.044329,-0.387688,-0.950149,-1.244981,-2.246231,0.076095,...,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0
4574,1.728292,-1.178441,-0.013288,-0.822921,-0.044329,-0.387688,0.589233,1.104181,0.698048,-1.679531,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0
2082,-0.422862,0.680332,-1.406809,-0.363421,-1.452452,-0.777343,1.102360,-0.461927,1.701447,-0.093229,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
1157,1.728292,-0.846517,-1.499710,1.802794,-1.452452,0.001967,-0.950149,0.321127,-0.678879,-1.020057,...,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
957,1.011241,0.481178,0.172515,-0.625992,-0.494928,1.950242,-0.437022,0.321127,0.478326,-0.004111,...,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0


* Part C: Bagging (Bootstrap Aggregating): 

Implement Bagging Classifier for course completion prediction.

In [7]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.metrics import accuracy_score, classification_report

dt_classifier = DecisionTreeClassifier(random_state=42)
bagging_classifier = BaggingClassifier(
    estimator=dt_classifier,
    n_estimators=100,
    random_state=42
)

bagging_classifier.fit(X_train_cls, y_train_cls)
y_pred_bag_cls = bagging_classifier.predict(X_test_cls)

print('Bagging Classifier Accuracy:', accuracy_score(y_test_cls, y_pred_bag_cls))
print('\nClassification Report:')
print(classification_report(y_test_cls, y_pred_bag_cls))

Bagging Classifier Accuracy: 0.7096153846153846

Classification Report:
              precision    recall  f1-score   support

           0       0.75      0.81      0.78       650
           1       0.63      0.55      0.59       390

    accuracy                           0.71      1040
   macro avg       0.69      0.68      0.68      1040
weighted avg       0.70      0.71      0.70      1040



Implement Bagging Regressor for final score prediction.

In [8]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

dt_regressor = DecisionTreeRegressor(random_state=42)

bagging_regressor = BaggingRegressor(
    estimator=dt_regressor,
    n_estimators=100,
    random_state=42
)

bagging_regressor.fit(X_train_reg, y_train_reg)
y_pred_bag_reg = bagging_regressor.predict(X_test_reg)
print('MAE :', mean_absolute_error(y_test_reg, y_pred_bag_reg))
print('MSE :', mean_squared_error(y_test_reg, y_pred_bag_reg))
print('RMSE:', mean_squared_error(y_test_reg, y_pred_bag_reg) ** 0.5)
print('R² Score:', r2_score(y_test_reg, y_pred_bag_reg))

MAE : 8.03628076923077
MSE : 100.71939372499999
RMSE: 10.035905226983761
R² Score: 0.46093112294214134


Compare bagging results with a single base model.

In [9]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

single_classifier = DecisionTreeClassifier(random_state=42)
single_classifier.fit(X_train_cls, y_train_cls)
y_pred_single_cls = single_classifier.predict(X_test_cls)

single_accuracy = accuracy_score(y_test_cls, y_pred_single_cls)
bagging_accuracy = accuracy_score(y_test_cls, y_pred_bag_cls)

from sklearn.metrics import r2_score

single_regressor = DecisionTreeRegressor(random_state=42)
single_regressor.fit(X_train_reg, y_train_reg)
y_pred_single_reg = single_regressor.predict(X_test_reg)

single_r2 = r2_score(y_test_reg, y_pred_single_reg)
bagging_r2 = r2_score(y_test_reg, y_pred_bag_reg)

In [10]:
comparison = pd.DataFrame({
    "Model": ['Decision Tree Classifier','Bagging Classifier','Decision Tree Regressor','Bagging Regressor'],
    "Metric": ['Accuracy','Accuracy','R² Score','R² Score'],
    "Score": [single_accuracy,bagging_accuracy,single_r2,bagging_r2]
})

comparison

,Model,Metric,Score
0,Decision Tree Classifier,Accuracy,0.628846
1,Bagging Classifier,Accuracy,0.709615
2,Decision Tree Regressor,R² Score,-0.106316
3,Bagging Regressor,R² Score,0.460931


* Part D: Boosting Algorithms 

Implement AdaBoost Classifier

In [11]:
from sklearn.ensemble import AdaBoostClassifier

ada_classifier = AdaBoostClassifier(
    n_estimators=100,
    learning_rate=1.0,
    random_state=42
)

ada_classifier.fit(X_train_cls, y_train_cls)
y_pred_ada_cls = ada_classifier.predict(X_test_cls)

print("Accuracy:", accuracy_score(y_test_cls, y_pred_ada_cls))
print("\nClassification Report:")
print(classification_report(y_test_cls, y_pred_ada_cls))

C:\Users\adity\AppData\Roaming\Python\Python312\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


Accuracy: 0.7221153846153846

Classification Report:
              precision    recall  f1-score   support

           0       0.75      0.82      0.79       650
           1       0.65      0.55      0.60       390

    accuracy                           0.72      1040
   macro avg       0.70      0.69      0.69      1040
weighted avg       0.72      0.72      0.72      1040



Implement AdaBoost Regressor

In [12]:
from sklearn.ensemble import AdaBoostRegressor

ada_regressor = AdaBoostRegressor(
    n_estimators=100,
    learning_rate=1.0,
    random_state=42
)

ada_regressor.fit(X_train_reg, y_train_reg)
y_pred_ada_reg = ada_regressor.predict(X_test_reg)

print("MAE :", mean_absolute_error(y_test_reg, y_pred_ada_reg))
print("MSE :", mean_squared_error(y_test_reg, y_pred_ada_reg))
print("RMSE:", mean_squared_error(y_test_reg, y_pred_ada_reg, squared=False))
print("R² Score:", r2_score(y_test_reg, y_pred_ada_reg))

MAE : 8.567658928247528
MSE : 110.14202555118264
RMSE: 10.494857100083957
R² Score: 0.41049944966044316


C:\Users\adity\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


Implement Gradient Boosting Classifier

In [13]:
from sklearn.ensemble import GradientBoostingClassifier

gb_classifier = GradientBoostingClassifier(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

gb_classifier.fit(X_train_cls, y_train_cls)
y_pred_gb_cls = gb_classifier.predict(X_test_cls)

print("Accuracy:", accuracy_score(y_test_cls, y_pred_gb_cls))
print("\nClassification Report:")
print(classification_report(y_test_cls, y_pred_gb_cls))

Accuracy: 0.7384615384615385

Classification Report:
              precision    recall  f1-score   support

           0       0.77      0.83      0.80       650
           1       0.68      0.58      0.63       390

    accuracy                           0.74      1040
   macro avg       0.72      0.71      0.71      1040
weighted avg       0.73      0.74      0.73      1040



Implement Gradient Boosting Regressor

In [14]:
from sklearn.ensemble import GradientBoostingRegressor

gb_regressor = GradientBoostingRegressor(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

gb_regressor.fit(X_train_reg, y_train_reg)
y_pred_gb_reg = gb_regressor.predict(X_test_reg)

print("MAE :", mean_absolute_error(y_test_reg, y_pred_gb_reg))
print("MSE :", mean_squared_error(y_test_reg, y_pred_gb_reg))
print("RMSE:", mean_squared_error(y_test_reg, y_pred_gb_reg, squared=False))
print("R² Score:", r2_score(y_test_reg, y_pred_gb_reg))

MAE : 7.858580366498315
MSE : 96.21571643066729
RMSE: 9.808961027074544
R² Score: 0.4850356391817412


C:\Users\adity\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


Implement LightGBM Classifier

In [15]:
from lightgbm import LGBMClassifier

lgbm_classifier = LGBMClassifier(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

lgbm_classifier.fit(X_train_cls, y_train_cls)
y_pred_lgbm_cls = lgbm_classifier.predict(X_test_cls)

print("Accuracy:", accuracy_score(y_test_cls, y_pred_lgbm_cls))
print("\nClassification Report:")
print(classification_report(y_test_cls, y_pred_lgbm_cls))

[LightGBM] [Info] Number of positive: 1562, number of negative: 2598
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.003402 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1075
[LightGBM] [Info] Number of data points in the train set: 4160, number of used features: 30
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.375481 -> initscore=-0.508775
[LightGBM] [Info] Start training from score -0.508775
Accuracy: 0.7259615384615384

Classification Report:
              precision    recall  f1-score   support

           0       0.76      0.82      0.79       650
           1       0.65      0.57      0.61       390

    accuracy                           0.73      1040
   macro avg       0.71      0.70      0.70      1040
weighted avg       0.72      0.73      0.72      1040



Implement LightGBM Regressor

In [16]:
from lightgbm import LGBMRegressor

lgbm_regressor = LGBMRegressor(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

lgbm_regressor.fit(X_train_reg, y_train_reg)
y_pred_lgbm_reg = lgbm_regressor.predict(X_test_reg)

print("MAE :", mean_absolute_error(y_test_reg, y_pred_lgbm_reg))
print("MSE :", mean_squared_error(y_test_reg, y_pred_lgbm_reg))
print("RMSE:", mean_squared_error(y_test_reg, y_pred_lgbm_reg, squared=False))
print("R² Score:", r2_score(y_test_reg, y_pred_lgbm_reg))

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000590 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1074
[LightGBM] [Info] Number of data points in the train set: 4160, number of used features: 30
[LightGBM] [Info] Start training from score 74.774183
MAE : 8.03153250197581
MSE : 100.45720486786034
RMSE: 10.02283417341923
R² Score: 0.4623344063374054


C:\Users\adity\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


Implement XGBoost Classifier

In [17]:
from xgboost import XGBClassifier

xgb_classifier = XGBClassifier(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42,
    eval_metric="logloss"
)
xgb_classifier.fit(X_train_cls, y_train_cls)
y_pred_xgb_cls = xgb_classifier.predict(X_test_cls)

print("Accuracy:", accuracy_score(y_test_cls, y_pred_xgb_cls))
print("\nClassification Report:")
print(classification_report(y_test_cls, y_pred_xgb_cls))

Accuracy: 0.7230769230769231

Classification Report:
              precision    recall  f1-score   support

           0       0.76      0.81      0.79       650
           1       0.65      0.58      0.61       390

    accuracy                           0.72      1040
   macro avg       0.70      0.69      0.70      1040
weighted avg       0.72      0.72      0.72      1040



Implement XGBoost Regressor

In [18]:
from xgboost import XGBRegressor

xgb_regressor = XGBRegressor(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

xgb_regressor.fit(X_train_reg, y_train_reg)
y_pred_xgb_reg = xgb_regressor.predict(X_test_reg)

print("MAE :", mean_absolute_error(y_test_reg, y_pred_xgb_reg))
print("MSE :", mean_squared_error(y_test_reg, y_pred_xgb_reg))
print("RMSE:", mean_squared_error(y_test_reg, y_pred_xgb_reg, squared=False))
print("R² Score:", r2_score(y_test_reg, y_pred_xgb_reg))

MAE : 8.010089941024779
MSE : 99.20402115948072
RMSE: 9.960121543409032
R² Score: 0.4690416779902502


C:\Users\adity\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


Compare performance and robustness with other boosting models.

In [19]:
comparison = pd.DataFrame({
    "Model": ["AdaBoost","Gradient Boosting","LightGBM","XGBoost"],
    "Classification Accuracy": [
        accuracy_score(y_test_cls, y_pred_ada_cls),
        accuracy_score(y_test_cls, y_pred_gb_cls),
        accuracy_score(y_test_cls, y_pred_lgbm_cls),
        accuracy_score(y_test_cls, y_pred_xgb_cls)
    ],
    "Regression R² Score": [
        r2_score(y_test_reg, y_pred_ada_reg),
        r2_score(y_test_reg, y_pred_gb_reg),
        r2_score(y_test_reg, y_pred_lgbm_reg),
        r2_score(y_test_reg, y_pred_xgb_reg)
    ]
})

comparison

,Model,Classification Accuracy,Regression R² Score
0,AdaBoost,0.722115,0.410499
1,Gradient Boosting,0.738462,0.485036
2,LightGBM,0.725962,0.462334
3,XGBoost,0.723077,0.469042


Part E: Voting & Stacking Ensembles

Implement Voting Classifier using multiple base classifiers.

In [20]:
from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

dt = DecisionTreeClassifier(random_state=42)
lr = LogisticRegression(max_iter=1000, random_state=42)
knn = KNeighborsClassifier()

hard_voting = VotingClassifier(
    estimators=[
        ('dt', dt),
        ('lr', lr),
        ('knn', knn)
    ],
    voting='hard'
)

hard_voting.fit(X_train_cls, y_train_cls)
y_pred_hard = hard_voting.predict(X_test_cls)
print("Hard Voting Accuracy:", accuracy_score(y_test_cls, y_pred_hard))

Hard Voting Accuracy: 0.7240384615384615


Compare Hard Voting vs Soft Voting.

In [21]:
soft_voting = VotingClassifier(
    estimators=[
        ('dt', dt),
        ('lr', lr),
        ('knn', knn)
    ],
    voting='soft'
)

soft_voting.fit(X_train_cls, y_train_cls)
y_pred_soft = soft_voting.predict(X_test_cls)

print('Hard Voting Accuracy :', accuracy_score(y_test_cls, y_pred_hard))
print('Soft Voting Accuracy :', accuracy_score(y_test_cls, y_pred_soft))

Hard Voting Accuracy : 0.7240384615384615
Soft Voting Accuracy : 0.6730769230769231


Implement Stacking Classifier with a meta-learner.

In [22]:
from sklearn.ensemble import StackingClassifier

stack_classifier = StackingClassifier(
    estimators=[
        ('dt', DecisionTreeClassifier(random_state=42)),
        ('knn', KNeighborsClassifier()),
        ('lr', LogisticRegression(max_iter=1000))
    ],
    final_estimator=LogisticRegression()
)

stack_classifier.fit(X_train_cls, y_train_cls)
y_pred_stack_cls = stack_classifier.predict(X_test_cls)
print("Stacking Accuracy:",accuracy_score(y_test_cls, y_pred_stack_cls))

Stacking Accuracy: 0.7432692307692308


Implement Stacking Regressor for score prediction.

In [23]:
from sklearn.ensemble import StackingRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

stack_regressor = StackingRegressor(
    estimators=[
        ('dt', DecisionTreeRegressor(random_state=42)),
        ('knn', KNeighborsRegressor()),
        ('lr', LinearRegression())
    ],
    final_estimator=LinearRegression()
)

stack_regressor.fit(X_train_reg, y_train_reg)
y_pred_stack_reg = stack_regressor.predict(X_test_reg)
print("MAE :", mean_absolute_error(y_test_reg, y_pred_stack_reg))
print("RMSE:", mean_squared_error(y_test_reg, y_pred_stack_reg)**0.5)
print("R² Score:", r2_score(y_test_reg, y_pred_stack_reg))

MAE : 7.81984175576198
RMSE: 9.731519119178975
R² Score: 0.4931348449109083


Part F: Model Evaluation & Comparison

Evaluate classification models using:
Accuracy, Precision, Recall, F1-score, ROC-AUC

In [24]:
from sklearn.metrics import precision_score,recall_score,f1_score,roc_auc_score

print('Accuracy :', accuracy_score(y_test_cls, y_pred_stack_cls))
print('Precision:', precision_score(y_test_cls, y_pred_stack_cls))
print('Recall   :', recall_score(y_test_cls, y_pred_stack_cls))
print('F1 Score :', f1_score(y_test_cls, y_pred_stack_cls))
print('ROC-AUC  :', roc_auc_score(y_test_cls, y_pred_stack_cls))

Accuracy : 0.7432692307692308
Precision: 0.6869300911854104
Recall   : 0.5794871794871795
F1 Score : 0.6286509040333796
ROC-AUC  : 0.7105128205128206


Evaluate regression models using:
MAE, RMSE, R² Score

In [25]:
print("MAE :", mean_absolute_error(y_test_reg, y_pred_stack_reg))
print("RMSE:", mean_squared_error(y_test_reg, y_pred_stack_reg)**0.5)
print("R² Score:", r2_score(y_test_reg, y_pred_stack_reg))

MAE : 7.81984175576198
RMSE: 9.731519119178975
R² Score: 0.4931348449109083


Compare all ensemble techniques and identify the best performing model.

Classification Comparison

In [26]:
classification_results = pd.DataFrame({
    "Model":["Bagging","AdaBoost","Gradient Boosting","LightGBM","XGBoost","Hard Voting","Soft Voting","Stacking"],
    "Accuracy":[
        accuracy_score(y_test_cls, y_pred_bag_cls),
        accuracy_score(y_test_cls, y_pred_ada_cls),
        accuracy_score(y_test_cls, y_pred_gb_cls),
        accuracy_score(y_test_cls, y_pred_lgbm_cls),
        accuracy_score(y_test_cls, y_pred_xgb_cls),
        accuracy_score(y_test_cls, y_pred_hard),
        accuracy_score(y_test_cls, y_pred_soft),
        accuracy_score(y_test_cls, y_pred_stack_cls)
    ]
})

classification_results.sort_values(
    by="Accuracy",
    ascending=False
)

,Model,Accuracy
7,Stacking,0.743269
2,Gradient Boosting,0.738462
3,LightGBM,0.725962
5,Hard Voting,0.724038
4,XGBoost,0.723077
1,AdaBoost,0.722115
0,Bagging,0.709615
6,Soft Voting,0.673077


Regression Comparison

In [27]:
regression_results = pd.DataFrame({
    "Model":["Bagging","AdaBoost","Gradient Boosting","LightGBM","XGBoost","Stacking"],
    "R² Score":[
        r2_score(y_test_reg, y_pred_bag_reg),
        r2_score(y_test_reg, y_pred_ada_reg),
        r2_score(y_test_reg, y_pred_gb_reg),
        r2_score(y_test_reg, y_pred_lgbm_reg),
        r2_score(y_test_reg, y_pred_xgb_reg),
        r2_score(y_test_reg, y_pred_stack_reg)
    ]
})

regression_results.sort_values(
    by="R² Score",
    ascending=False
)

,Model,R² Score
5,Stacking,0.493135
2,Gradient Boosting,0.485036
4,XGBoost,0.469042
3,LightGBM,0.462334
0,Bagging,0.460931
1,AdaBoost,0.410499


Best Performing Model: After running the comparison tables, Stacking has the highest Accuracy - 0.743269 for classification and the highest R² Score: Stacking - 0.493135 for regression.

* Part G: Final Analysis & Reporting

-1. Impact of Bagging vs Boosting

Bagging (Bootstrap Aggregating):

-Bagging trains multiple models independently using different bootstrap samples of the training data. 

-The final prediction is obtained by averaging (regression) or majority voting (classification). 

-It mainly reduces model variance and helps prevent overfitting. 

-Bagging is easy to parallelize because all base models are trained independently. 

-Example: Bagging Classifier and Bagging Regressor.

Boosting:

-Boosting trains models sequentially, where each new model focuses on correcting the errors made by previous models. 

-It reduces both bias and variance, often achieving higher predictive accuracy. 

-Boosting generally provides better performance than bagging but requires more training time. 

-Examples: AdaBoost, Gradient Boosting, LightGBM, and XGBoost.

-2. Comparison of Tree-Based Boosting Algorithms

- AdaBoost:

-Uses weak learners sequentially. 

-Increases the weight of misclassified samples. 

-Performs well on relatively simple datasets.
Sensitive to noisy data and outliers.

- Gradient Boosting:

-Builds trees sequentially by minimizing prediction errors. 

-Produces higher accuracy than AdaBoost. 

-Requires careful tuning of hyperparameters. 

- LightGBM:

-Uses histogram-based tree learning. 

-Faster training and lower memory consumption. 

-Efficient for large datasets with many features. 

-Handles categorical features effectively.

- XGBoost:

-Includes regularization to reduce overfitting. 

-Handles missing values automatically. 

-Highly optimized for speed and accuracy.

-Generally achieves the best predictive performance among boosting algorithms.

-3. Advantages of Voting and Stacking 

- Voting Ensemble: 

-Combines predictions from multiple classifiers. 

-Reduces individual model errors. 

-Improves prediction stability. 

-Soft Voting generally performs better than Hard Voting because it considers prediction probabilities.

- Stacking Ensemble: 

-Combines predictions from several base models using a meta-learner. 

-Learns how to best combine different models. 

-Usually provides higher accuracy than individual models. 

-Reduces the weaknesses of single algorithms.

-4. Final Model Recommendation for Deployment

For Classification: In most cases, XGBoost Classifier or Stacking Classifier provides the best overall performance.

Recommendation: highest accuracy obtained Stacking.

For Regression: In most cases, XGBoost Regressor or Stacking Regressor achieves the best predictive performance. 

Recommendation: highest R² score and lowest error obtained in Stacking